### PR TITLE
Add Brain MRI Super-Resolution Network (ESPCN)

### PR DESCRIPTION
**Pull request by Benjamin-Guy**
This pull-request introduces an implementation of the Efficient Sub-Pixel CNN (ESPCN) model used as the brain MRI super-resolution network for 4x up-scaling images from the ADNI brain dataset. The model was tested on 4x down-scaled images from the ADNI brain dataset, in which the model used 4x up-scaling to bring the image back to the original resolution. Perceptual quality of the up-scaled images in comparison with the original images shows a reasonably clear image and good results. Here is an example of the perceptual quality of up-scaled images produced by the model:

Figure_1

Changes made:
- Add dataset.py which preprocesses images from the ADNI brain dataset and sets up train, validation, and test data loaders.
- Add modules.py which contains the ESPCN model.
- Add train.py which trains the ESPCN model using the ADNI brain dataset train and validation data loaders.
- Add predict.py which uses the test data loader to produce plots of the original, down-scaled, and up-scaled images for comparison. 
- Add images folder which stores images that are used for documentation.

Ensure that PyTorch (version 2.1.0+cu121) and PyTorch Vision (version 0.16.0+cu121) are installed. Whilst not a requirement, it is recommended that a GPU is available for training the model as it will greatly decrease the time taken to train the model. The ADNI brain dataset will need to be downloaded externally to the project and paths to the dataset will need to be set in the dataset.py file. 

Author: Benjamin Guy (46972990)

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Overall nice implementation of PixelCNN, only issue to images is the reconstructed image contrast is different to target image. Correctly tracks training and validation losses to demonstrate there is no over-fitting. End of training loop will report average MSE over the test set (though "predict.py" will only show some input/output images). 

**[Recognition Problem]**
"modules.py":

- Implements PixelCNN without any architectural changes. Does allow for a variable upscaling factor. All looks ok.

"dataset.py":

- SuperResolutionDataset has a redundant resize, otherwise the implementation is nice and makes use of ImageFolder.
- Creates a train and validation set randomly.

"train.py":

- Nice training loop, correctly aggregates and reports training and validation losses at the end of each epoch. 
- Provides a mechanism to get average test loss on the test dataset at the end of training.
- Saves model correctly and generates some figures. Though simply saves the model at the end of the final epoch rather than checking for lowest validation loss.

"predict.py":

- Doesn't provide a way to see metrics on the test dataset (but this is in "train.py").
- Correctly loads trained weights and generates some input/output figures.

"README.MD":

- Provides a description of the network architecture and prints the network structure.
- Provides some examples of the inputs and outputs, results seem ok but with a slight change to the contrast of the images.
- Provides the folder structure (same as provided as the dataloaders handle the AD and NC folders).
- Provides a training and validation loss to demonstrate there is no over-fitting occurring, however, it does seem the network could have continued training for a few more epochs (only 10 run).
- No dependencies noted.

**[Commit Log]**
Seems to have a thorough commit log.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
